### PR TITLE
AI detection multitools can show camera static and AI eyes

### PR DESCRIPTION
**Pull request by Cruix**
:cl:
add: AI detection multitools can now be toggled to provide a HUD that shows the exact viewports of AI eyes, and the location of nearby camera blind spots.
/:cl:

Artist's rendering:
39848978-142cddf8-53d8-11e8-8d00-51dec55605b7

The old color-change functionality of the multitool remains unchanged. The hud is toggled with an action button, because multitools already have an attack_self() that does something. When toggled on, you get the hud as long as the multitool is in any slot on your body.

The only way I could find to reliably show exactly where an AI is looking was to place an image on every turf they are looking at. Because AI eyes move around a lot and I do not want them placing images all over the place when they do not have to, if there is nobody in the world currently using an ai detection multitool, they will not update the images until someone is.

Only AI eyes and multicamera zones currently appear. Advanced camera consoles do not appear on the hud, and no longer trigger the proximity color change. Also fixed multitools not appearing in-hands if their icon state changed.

Example screenshot:
untitled
The AI eye is the red overlay on the left, a 4x4 multicamera area is the orange overlay on the right, and the transparent static in the center is solid static to the AI.

### CODE REVIEW

**Comment by cacogen** [comment]
fantastic

**Comment by davethwave** [comment]
A nice buff that's reasonable.

**Comment by Spaceinaba** [comment]
I am an AI player and I approve this PR.

**Comment by zxaber** [comment]
Just to be sure, the static overlay updates with borgs too, right?

**Comment by YakumoChen** [comment]
Damn this is some cool spy shit, I can just hear spy music in my head looking at this and trying to sneak around in camera blind spots

**Comment by Ratvar** [comment]
Maybe make it into glasses if that's a hud?

**Comment by Dorsisdwarf** [comment]
>making the ai detector into a highly-useful tool
m-madness

**Comment by SpaceManiac** [comment]
Merge conflict.

Also what's your opinion on action button versus, say, alt-clicking? The item is supposed to be stealthy, after all. Otherwise very useful.

**Comment by Cruix** [comment]
The kind of people who will pick up the multitool to see if it has an action are the kind of people who would find out anyway.

**Comment by Cruix** [comment]
Making the static unclickable will require potentially doubling the already obscene number of images that the camera chunk system already uses.

**Comment by MrDoomBringer** [comment]
>not giving credit for my artists rendering

reeeeeeeeee

**Comment by SpaceManiac** [comment]
The static is working great in testing! Do you think you'd be able to adapt the red and orange areas to `vis_contents` for the clickthrough as well?

**Comment by Cruix** [comment]
I am working on that. I would like to do it slightly differently, because they use of datum huds.

**Comment by Cyberboss** [comment]
Where do we stand on this?

**Comment by Cruix** [comment]
Waiting for #38754 to be merged or no longer testmerged.

**Comment by optimumtact** [comment]
@Cruix ready?

**Comment by Cruix** [comment]
Ready

File: code/game/objects/items/devices/multitool.dm [hunk 1]
```diff
@@ -98,30 +97,89 @@
...
 
 /obj/item/multitool/ai_detect/Destroy()
 	STOP_PROCESSING(SSobj, src)
+	if(hud_on && ismob(loc))
+		remove_hud(loc)
+	QDEL_NULL(toggle_action)
+	QDEL_NULL(eye)
 	return ..()
 
+/obj/item/multitool/ai_detect/ui_action_click()
```
Comment (Cyberboss) [inline code/game/objects/items/devices/multitool.dm]: We don't proliferate ui_action_click anymore. Use a custom action datum

File: code/modules/mob/living/silicon/ai/freelook/cameranet.dm [hunk 2]
```diff
@@ -18,15 +18,24 @@ GLOBAL_DATUM_INIT(cameranet, /datum/cameranet, new)
 	// The object used for the clickable stat() button.
 	var/obj/effect/statclick/statclick
 
-	// The object used in vis_contents of obscured turfs
-	var/vis_contents
+	// The objects used in vis_contents of obscured turfs
+	var/list/vis_contents
```
Comment (Cyberboss) [inline code/modules/mob/living/silicon/ai/freelook/cameranet.dm]: can you name this something else so as not to cause confusion please

File: code/modules/mob/living/silicon/ai/freelook/eye.dm [hunk 3]
```diff
@@ -47,18 +100,18 @@
...
 		ai.all_eyes -= src
 		ai = null
 	for(var/V in visibleCameraChunks)
 		var/datum/camerachunk/c = V
 		c.remove(src)
+	GLOB.aiEyes -= src
+	if(ai_detector_visible)
+		var/datum/atom_hud/ai_detector/hud = GLOB.huds[DATA_HUD_AI_DETECT]
+		hud.remove_from_hud(src)
+		QDEL_LIST(hud_list[AI_DETECT_HUD])
```
Comment (Cyberboss) [inline code/modules/mob/living/silicon/ai/freelook/eye.dm]: This will resolve to `hud_list[AI_DETECT_HUD].Cut()` which is actually a : operator. No dice

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/datums/holocall.dm [hunk 4]
```diff
@@ -60,7 +60,6 @@
 		user.remote_control = null
 
 	if(!QDELETED(eye))
-		eye.RemoveImages()
```
[the remark this answers is not in the document]
Comment (Cruix) [reply code/datums/holocall.dm]: This was replaced by code in the eye object's Destroy()

File: code/game/objects/items/devices/multitool.dm (hunk 1 above)
Comment (Cruix) [reply code/game/objects/items/devices/multitool.dm]: I did use a custom action datum, this is here to suppress the default behavior of ui_action_click().

File: code/modules/mob/living/silicon/ai/freelook/eye.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/mob/living/silicon/ai/freelook/eye.dm]: It's pretty silly to have to worry about implicit : operators in the expansion of defines

**Comment by optimumtact** [comment]
@Cruix conflicting unfortunately, the rest is fine though (despite the stray map diff failure)

**Pull request merged**